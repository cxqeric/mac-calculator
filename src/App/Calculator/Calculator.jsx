import React from 'react';
import PropTypes from 'prop-types';

import {
  START,
  START_DECIMAL,
  CUMUL,
  CUMUL_DECIMAL,
  LOW,
  availableShortcuts,
  cumulate,
  multiply,
  getDataByValue,
  getDataByShortcut } from '../../utils';

const initialState = {
  inputs: [0],
  pendingOps: [],
  target: {
    index: 0,
    status: START
  },
  output: '0',
  initial: true
};

class Calculator extends React.Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ])
  };

  static defaultProps = {
    children: null
  };

  static CALCULATOR_CONTEXT = '__calculator__';

  static childContextTypes = {
    [Calculator.CALCULATOR_CONTEXT]: PropTypes.shape({
      initial: PropTypes.bool,
      output: PropTypes.string,
      handleInput: PropTypes.func
    })
  };

  state = { ...initialState };


  getChildContext = () => ({
    [Calculator.CALCULATOR_CONTEXT]: {
      initial: this.state.initial,
      output: this.state.output,
      handleInput: this.handleInput
    }
  });

  componentDidMount() {
    document.onkeydown = this.handleKeyPress;
  }


  handleKeyPress = ({ key: shortcut }) => {
    if (!availableShortcuts.includes(shortcut)) {
      return;
    }
    const { value } = getDataByShortcut({ shortcut });
    this.handleInput({ value });
  }

  handleInput = ({ value: targetValue }) => {
    const targetData = getDataByValue({ value: targetValue });
    return targetData ?
      this[`handle${targetData.type}`](targetData) :
      null;
  };

  handleDigit = ({ value: digit }) => {
    const target = { ...this.state.target };
    const inputs = [...this.state.inputs];
    const { index, status } = { ...target };
    const output = cumulate(inputs[index] || 0, digit, status);
    if (status === START) {
      target.status = CUMUL;
      if (index === 2 && typeof inputs[2] !== 'undefined') {
        Object.assign(inputs, { 0: inputs[1], 1: inputs[0] });
      }
    } else if (status === START_DECIMAL) {
      target.status = CUMUL_DECIMAL;
    }
    inputs[index] = output;
    this.setState({
      inputs, target, output, initial: false
    });
  };

  handleBinary = (operation) => {
    const target = { ...this.state.target };
    const pendingOps = [...this.state.pendingOps];
    const { index } = { ...target };
    if (index === 0) {
      target.status = START;
      target.index = 1;
      pendingOps[0] = operation;
      return this.setState({
        pendingOps,
        target
      });
    }

    return operation.priority === LOW ?
      this.inputLowBinaryOp(operation) :
      this.inputHighBinaryOp(operation);
  };

  inputLowBinaryOp = (operation) => {
    const target = { ...this.state.target };
    let inputs = [...this.state.inputs];
    let pendingOps = [...this.state.pendingOps];
    const { index } = { ...target };
    let { output } = { ...this.state };
    target.status = START;
    if (index === 1) {
      if (typeof inputs[1] !== 'undefined') {
        output = pendingOps[0].work(inputs[0], inputs[1]);
      }
      pendingOps[0] = operation;
    } else if (typeof inputs[2] !== 'undefined') {
      output = pendingOps[0].work(inputs[0], pendingOps[1].work(inputs[1], inputs[2]));
    } else {
      output = pendingOps[0].work(inputs[0], inputs[1]);
      inputs[1] = output;
      pendingOps = [operation];
    }
    inputs = [output];
    target.index = 1;
    this.setState({
      inputs,
      target,
      pendingOps,
      output,
      initial: false
    });
  };

  inputHighBinaryOp = (operation) => {
    const target = { ...this.state.target };
    let inputs = [...this.state.inputs];
    const pendingOps = [...this.state.pendingOps];
    const { index } = { ...target };
    let { output } = { ...this.state };
    target.status = START;
    if (typeof inputs[index] === 'undefined') {
      pendingOps[index - 1] = operation;
    } else if (index === 1) {
      if (pendingOps[0].priority === LOW) {
        pendingOps[1] = operation;
        target.index = 2;
      } else {
        output = pendingOps[0].work(inputs[0], inputs[1]);
        inputs = [output];
        pendingOps[0] = operation;
        target.index = 0;
      }
    } else {
      output = pendingOps[1].work(inputs[1], inputs[2]);
    }
    this.setState({
      inputs,
      target,
      pendingOps,
      output,
      initial: false
    });
  };

  handleUnary = ({ key, work }) => {
    const target = { ...this.state.target };
    const inputs = [...this.state.inputs];
    const { index, status } = { ...target };
    let { output } = { ...this.state };
    if (index === 0 || status !== START) {
      output = work(inputs[index]);
      target.status = START;
    } else {
      output = key === '%' ?
        multiply(inputs[index - 1], work(inputs[index - 1])) :
        work(inputs[index - 1]);
      if (index === 1) {
        target.index = 0;
      }
    }
    inputs[index] = output;
    this.setState({
      inputs, target, output, initial: false
    });
  };

  handleDot = () => {
    const target = { ...this.state.target };
    const inputs = [...this.state.inputs];
    const { index, status } = { ...target };
    if (status === CUMUL_DECIMAL) {
      return;
    }
    target.status = START_DECIMAL;
    if (typeof inputs[index] === 'undefined') {
      inputs[index] = 0;
    }
    const output = `${inputs[index]}.`;
    this.setState({
      inputs, target, output, initial: false
    });
  };

  handleEqual = () => {
    const target = { ...this.state.target };
    let inputs = [...this.state.inputs];
    let pendingOps = [...this.state.pendingOps];
    const { index } = { ...target };
    let { output } = { ...this.state };

    target.status = START;
    switch (index) {
      case 0:
        if (typeof pendingOps[0] !== 'undefined') {
          output = pendingOps[0].work(inputs[0], inputs[1]);
        }
        break;
      case 1:
        if (typeof inputs[1] === 'undefined') {
          inputs = [inputs[0], inputs[0]];
        }
        output = pendingOps[0].work(inputs[0], inputs[1]);
        break;
      default:
        if (typeof inputs[2] !== 'undefined') {
          output = pendingOps[0].work(inputs[0], pendingOps[1].work(inputs[1], inputs[2]));
          inputs = [inputs[0], inputs[2]];
        } else {
          output = pendingOps[0].work(pendingOps[1].work(inputs[0], inputs[1]), inputs[1]);
        }
        pendingOps = [pendingOps[1]];
    }
    inputs[0] = output;
    target.index = 0;
    this.setState({
      inputs,
      target,
      output,
      pendingOps
    });
  };

  handleReset = () => this.setState(initialState);

  render() {
    return (
      <div className="calculator">
        { this.props.children }
      </div>
    );
  }
}


export default Calculator;
