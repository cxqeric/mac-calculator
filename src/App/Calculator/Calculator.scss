.calculator {
  display: flex;
  flex-flow: column nowrap;
  justify-content: flex-start;
  width: 100%;
  height: 100%;
  color: white;

  .buttons {
    flex-basis: 80%;
    display: flex;
    flex-wrap: wrap;
    width: 100%;
  }  
}