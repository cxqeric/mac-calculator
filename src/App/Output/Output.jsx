import React from 'react';
import PropTypes from 'prop-types';


class Output extends React.Component {
  static contextTypes = {
    output: PropTypes.string
  };

  state = {
    scale: 1
  }

  componentDidUpdate() {
    const scale = this.node.parentNode.offsetWidth / this.node.offsetWidth;
    if (scale === this.state.scale) {
      return;
    }
    if (scale < 1) {
      this.setState({ scale });
    } else if (this.state.scale < 1) {
      this.setState({ scale: 1 });
    }
  }

  refCallback = (node) => {
    this.node = node;
  }

  render() {
    const outputTextStyle = { transform: `scale(${this.state.scale})` };
    return (
      <div className="output">
        <div
          className="output-text"
          type="text"
          ref={this.refCallback}
          style={outputTextStyle}
        >
          {this.context.output}
        </div>
      </div>
    );
  }
}

export default Output;
