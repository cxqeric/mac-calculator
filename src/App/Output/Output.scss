.output {
  position: relative;
  grid-column: span 4;
  background: #111111;

  .output-text {
    position: absolute;
    right: 0;
    height: 100%;
    display: grid;
    align-content: end;
    padding: 0 5%;
    box-sizing: border-box;
    transform-origin: right;
    white-space: nowrap;
    font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
    font-size: 17em;
    font-weight: lighter;
  }
}