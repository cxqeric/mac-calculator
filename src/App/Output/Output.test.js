import React from 'react';
import { shallow } from 'enzyme';

import Calculator from '../Calculator/Calculator';
import Output from './Output';


describe('Output', () => {
  test('renders output', () => {
    const context = { [Calculator.CALCULATOR_CONTEXT]: { output: '0' } };
    const wrapper = shallow(<Output />, { context });
    expect(wrapper).toMatchSnapshot();
  });
});
