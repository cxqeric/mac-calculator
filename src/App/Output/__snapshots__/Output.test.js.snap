// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Output renders output 1`] = `
<div
  className="output"
>
  <div
    className="output-text"
    style={
      Object {
        "transform": "scale(1)",
      }
    }
    type="text"
  >
    0
  </div>
</div>
`;
